### PR TITLE
On session feature

### PR DESCRIPTION
**Pull request by vinaysurya**
## Description
This codeflow implements the OnSessions Feature For DotNet Core library. Also noticed some issues with the way we were doing close on our Client Entities, have addressed them as well.

Addresses:
https://github.com/Azure/azure-service-bus-dotnet/issues/35

This checklist is used to make sure that common guidelines for a pull request are followed.

- [X] **I have read the [contribution guidelines](./CONTRIBUTING.md).**
- [X] Title of the pull request is clear and informative.
- [X] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR.
- [X] The pull request does not introduce breaking changes (unless a major version change occurs in the assembly and module).
- [X] If applicable, the public code is properly documented.
- [X] Pull request includes test coverage for the included changes.
- [X ] The code builds without any errors.

### CODE REVIEW

File: src/Microsoft.Azure.ServiceBus/Amqp/AmqpLinkCreator.cs [hunk 1]
```diff
@@ -46,19 +46,33 @@ protected AmqpLinkCreator(string entityPath, ServiceBusConnection serviceBusConn
...
+            {
+                // Create Link
                 AmqpObject link = this.OnCreateAmqpLink(connection, this.amqpLinkSettings, session);
                 await link.OpenAsync(timeoutHelper.RemainingTime()).ConfigureAwait(false);
                 return link;
             }
-            catch (Exception)
+            catch (Exception exception)
             {
-                session?.Abort();
```
Comment (nemakam) [inline src/Microsoft.Azure.ServiceBus/Amqp/AmqpLinkCreator.cs]: Why remove this line?

File: src/Microsoft.Azure.ServiceBus/Core/MessageReceiver.cs [hunk 2]
```diff
@@ -379,13 +379,13 @@ public Task<IList<Message>> PeekAsync(int maxMessageCount)
...
         public void RegisterMessageHandler(Func<Message, CancellationToken, Task> handler)
         {
-            this.RegisterMessageHandler(handler, new RegisterHandlerOptions() { ReceiveTimeOut = this.OperationTimeout });
+            this.RegisterMessageHandler(handler, new RegisterMessageHandlerOptions());
         }
 
-        public void RegisterMessageHandler(Func<Message, CancellationToken, Task> handler, RegisterHandlerOptions registerHandlerOptions)
+        public void RegisterMessageHandler(Func<Message, CancellationToken, Task> handler, RegisterMessageHandlerOptions registerMessageHandlerOptions)
         {
-            registerHandlerOptions.ReceiveTimeOut = this.OperationTimeout;
```
Comment (nemakam) [inline src/Microsoft.Azure.ServiceBus/Core/MessageReceiver.cs]: if ReceiveTimeout is not specified in the options, then we will have to add it ourselves

File: src/Microsoft.Azure.ServiceBus/Amqp/AmqpLinkCreator.cs (hunk 1 above)
Comment (vinaysurya) [inline src/Microsoft.Azure.ServiceBus/Amqp/AmqpLinkCreator.cs]: I have not removed it. Just seperated the session creation/open from the link creation and not aborting the session when link creation fails

File: src/Microsoft.Azure.ServiceBus/Core/MessageReceiver.cs (hunk 2 above)
Comment (vinaysurya) [inline src/Microsoft.Azure.ServiceBus/Core/MessageReceiver.cs]: the default is initialized in the constructor for RegisterMessageHandlerOptions

**Pull request merged**